### PR TITLE
User story

### PR DESCRIPTION
**Pull request by helenyau0**

### CODE REVIEW

File: app/server.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+const app = express()
+
+require('dotenv').load();
+
+app.set('view engine', 'ejs')
+
+app.use(bodyParser.urlencoded({ extended: true }))
+app.use(bodyParser.json())
+
+app.use(cookieParser('secret'))
```
Comment (bundacia) [inline app/server.js]: you don't need cookie-parser middleware here. cookie-session will take care of that for you. So this line and the require on line 4 can be removed.

File: app/server.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+app.use(cookieSession({
+  secret : process.env.SECRET_KEY,
+  key    : ['key1', 'key2'],
```
Comment (bundacia) [inline app/server.js]: cookieSession needs either a `secret` OR `keys`. Since you have a secret here you can remove the `key` attribute from these options. Also note that your attribute is `key` here, for it to work it would need to be `keys` (plural). Also, the keys are sensitive and should not be checked in, so if you were going to use the `keys` version you'd need to store them in your .env file or something similar like you're doing for `SECRET_KEY`.

**tl;dr: you can delete line 20**

File: app/server.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+const loginRequired = (req, res, next) => {
+  if(!req.session.user) {
+    res.redirect('/login')
+  } else {
+    next()
+  }
+}
+
+app.get('/homepage', loginRequired, (req, res, next) => {
```
Comment (bundacia) [inline app/server.js]: When the specs reference the "homepage" they just mean `/`. Your `/` route should check for the user and show either "Welcome ${user}" or "Welcome stranger" depending on whether the user is logged in. The links shown should also be appropriate for the user's logged-in state (logout link if logged in, otherwise login/signup links, etc)

File: app/server.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  })
+})
+
+app.post('/signup', (req, res) => {
+  const { email, password, confirm_pass } = req.body
+
+  if(email.length <= 0 && password.length <= 0 && confirm_pass.length <= 0 || email.length < 3 && password.length < 3 && confirm_pass.length < 3) {
+    res.redirect('/flash')
+  } else if(password !== confirm_pass) {
+    res.redirect('/flash2')
```
Comment (bundacia) [inline app/server.js]: This error handling could be made quite a bit simpler. Instead of using multiple redirects and flash messages on an error consider doing this instead:

```js
} else if(password !== confirm_pass) {
  res.render('signup', { message: 'passwords do not match' })
}
```

And then just remove those `flash` routes altogether.

File: app/server.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    const saltRounds = 10
+
+    bcrypt.hash(password, saltRounds).then(function(hashedPassword) {
+      return hashedPassword
+    })
+    .then(function(hash) {
+      return db.addUsers(email, hash)
+    })
+    .then(result => {
```
Comment (bundacia) [inline app/server.js]: instead of calling this parameter `result` why not call it what it is, `user`?

File: app/server.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+      req.session.user = result
+      res.redirect('/homepage')
+    })
+    .catch(err => console.log(err))
```
Comment (bundacia) [inline app/server.js]: Since you're catching and logging the error, but not rendering anything to the user, the user's browser will just hang waiting for the server to respond (and probably eventually time out). You should add some error handling so that an error message gets displayed to the user.

One option would be to add a `res.send` or `res.render` to this catch like this:

```js
.catch(err => {
  console.log(err)
  res.status(500).send('ERROR: ' + err.message)
})
```

You can also just use the `next()` callback and pass it to the error to hand off to express's built-in error handler.

```js
// notice I've added a third arg (next) to this route function
app.post('/signup', (req, res, next) => { 

   // ... details removed for brevity =)

    bcrypt.hash(password, saltRounds).then(function(hashedPassword) {
      return hashedPassword
    })
    .then(function(hash) {
      return db.addUsers(email, hash)
    })
    .then(result => {
      req.session.user = result
      res.redirect('/homepage')
    })
    .catch(err => next(err)) // catch callback just calls `next` with the error.
  }
})
```

This method is nice becuase express with set the appropriate status code and render the error to the user and log it as well.

Checkout the [express error handling docs](http://expressjs.com/en/guide/error-handling.html) for more info.

File: database/db.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  database: 'http_auth'
+}
+
+const db = pgp(cn)
+
+const getUser = (email) => {
+  return db.one('SELECT * FROM users WHERE email = $1', [email])
+}
+
+const addUsers = (email, password) => {
```
Comment (bundacia) [inline database/db.js]: this function name is plural but only adds a single user. Maybe rename it `addUser`?

File: app/server.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        res.redirect('/homepage')
+      }
+    });
+  })
+})
+
+app.post('/signup', (req, res) => {
+  const { email, password, confirm_pass } = req.body
+
+  if(email.length <= 0 && password.length <= 0 && confirm_pass.length <= 0 || email.length < 3 && password.length < 3 && confirm_pass.length < 3) {
```
Comment (bundacia) [inline app/server.js]: This is checking to see if all fields are less than zero or ALL fields are less than three. So it will be false if one of the fields is blank but the others are long enough. Also, you're checking elsewhere that confirm_pass is the same as password, so you don't have to look at it here at all. I think you can do something both simpler and more correct:

```js
if(email.length < 3 || password.length < 3) {
 // ...
```

**Review by bundacia — COMMENTED** [review]
Great job overall! I've recommended a few changes in in-line comments.

I like how you separated the database logic out into it's own module. You might consider doing the same thing with the logic for hashing and checking the password. Imagine a function called `getValidUser(email, password)` that used the db method to fetch the user and then did the password comparison as well, returning the user if the password was valid and null otherwise. You could create another function called `saveNewUser(email, password)` that took the email and raw password and handles the hashing of the password and then used `db.addUser` internally to save the user to the database.

The benefit here would be that your route code in server.js would get a bit simpler as you would be abstracting away details about everything except the http parameter validation and response.

File: app/server.js (hunk 6 above)
Comment (helenyau0) [reply app/server.js]: the next() option is super smooth :D
